File: frontend/src/app/core/services/admin-announcement.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from './api.service';
import { pageParams } from '../../shared/utils/page-params.util';
import {
  Announcement,
  AnnouncementDetail,
  AnnouncementKpis,
  AnnouncementFilters,
  AnnouncementRequest,
  Page,
} from '../../models/admin-announcement.model';

/**
 * Platform-staff API client for the Announcements console: the announcement queue (filterable +
 * paginated), lifecycle KPIs, per-record detail (+ change history) and the gated mutations
 * (create / save / publish / end / archive / duplicate). All gate on ANNOUNCEMENT_MANAGE + audit-logged.
 */
@Injectable({ providedIn: 'root' })
export class AdminAnnouncementService {
  private readonly api = inject(ApiService);
  private readonly base = '/admin/announcements';

  list(filters: AnnouncementFilters = {}, page = 0, size = 10) {
    const params = pageParams(filters, page, size);
    return this.api.get<Page<Announcement>>(this.base, { params });
  }

  kpis() { return this.api.get<AnnouncementKpis>(`${this.base}/kpis`); }
  get(id: string) { return this.api.get<AnnouncementDetail>(`${this.base}/${id}`); }

  create(body: AnnouncementRequest) { return this.api.post<Announcement>(this.base, body); }
  update(id: string, body: AnnouncementRequest) { return this.api.put<Announcement>(`${this.base}/${id}`, body); }
  publish(id: string) { return this.api.post<Announcement>(`${this.base}/${id}/publish`, {}); }
  end(id: string) { return this.api.post<Announcement>(`${this.base}/${id}/end`, {}); }
  archive(id: string) { return this.api.post<Announcement>(`${this.base}/${id}/archive`, {}); }
  duplicate(id: string) { return this.api.post<Announcement>(`${this.base}/${id}/duplicate`, {}); }
}

File: backend/src/main/java/com/postwerk/service/AdminAnnouncementService.java
package com.postwerk.service;

import com.postwerk.dto.admin.AdminAnnouncementDetailResponse;
import com.postwerk.dto.admin.AdminAnnouncementResponse;
import com.postwerk.dto.admin.AnnouncementKpisResponse;
import com.postwerk.dto.admin.AnnouncementRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Platform-staff Announcements console: the announcement queue, lifecycle KPIs, per-record detail
 * (+ change history) and the gated mutations (create / save / publish / end / archive / duplicate).
 * Read + mutate both gate on {@code ANNOUNCEMENT_MANAGE}; every mutation is audit-logged.
 *
 * @since 1.0
 */
public interface AdminAnnouncementService {

    Page<AdminAnnouncementResponse> list(String search, String type, String status, String audience,
                                         String placement, String sort, Pageable pageable);

    AnnouncementKpisResponse kpis();

    AdminAnnouncementDetailResponse getAnnouncement(UUID id);

    AdminAnnouncementResponse create(AnnouncementRequest req, UUID actorUserId, String ip);

    AdminAnnouncementResponse update(UUID id, AnnouncementRequest req, UUID actorUserId, String ip);

    AdminAnnouncementResponse publish(UUID id, UUID actorUserId, String ip);

    AdminAnnouncementResponse end(UUID id, UUID actorUserId, String ip);

    AdminAnnouncementResponse archive(UUID id, UUID actorUserId, String ip);

    AdminAnnouncementResponse duplicate(UUID id, UUID actorUserId, String ip);
}
